app/core/Engine.ts

```diff
@@ -401,29 +399,33 @@ class Engine {
       return lowerCasedIdentities;
     };
 
-    const keyringState = initialKeyringState || initialState.KeyringController;
+    const qrKeyringBuilder = () => new QRHardwareKeyring();
+    qrKeyringBuilder.type = QRHardwareKeyring.type;
 
-    const keyringController = new KeyringController(
-      {
-        removeIdentity: preferencesController.removeIdentity.bind(
-          preferencesController,
-        ),
-        syncIdentities: preferencesController.syncIdentities.bind(
-          preferencesController,
-        ),
-        updateIdentities: preferencesController.updateIdentities.bind(
-          preferencesController,
-        ),
-        setSelectedAddress: preferencesController.setSelectedAddress.bind(
-          preferencesController,
-        ),
-        setAccountLabel: preferencesController.setAccountLabel.bind(
-          preferencesController,
-        ),
-      },
-      { encryptor, keyringTypes: additionalKeyrings },
-      keyringState,
-    );
+    const keyringController = new KeyringController({
+      removeIdentity: preferencesController.removeIdentity.bind(
+        preferencesController,
+      ),
+      syncIdentities: preferencesController.syncIdentities.bind(
+        preferencesController,
+      ),
+      updateIdentities: preferencesController.updateIdentities.bind(
+        preferencesController,
+      ),
+      setSelectedAddress: preferencesController.setSelectedAddress.bind(
+        preferencesController,
+      ),
+      setAccountLabel: preferencesController.setAccountLabel.bind(
+        preferencesController,
+      ),
+      encryptor,
+      // @ts-expect-error Error might be caused by base controller version mismatch
+      messenger: this.controllerMessenger.getRestricted({
+        name: 'KeyringController',
+      }),
```

Comment:
Events and Actions needed by the controller should be added here:

```suggestion
      messenger: this.controllerMessenger.getRestricted({
        name: 'KeyringController',
        allowedEvents: [
          'KeyringController:lock',
          'KeyringController:unlock',
          'KeyringController:stateChange',
          'KeyringController:accountRemoved',
        ],
        allowedActions: ['KeyringController:getState'],
      }),
```